### PR TITLE
APIs for retrieving models in parallel

### LINKED ISSUE #188: Threaded collection download with model dependencies causes errors
## Environment
* OS Version: Ubuntu 18.04
* Source and binary.
    
## Description

There is a race condition with threaded collection download that can cause console errors. The condition occurs when a collection contains multiple models that all depend on the same upstream model. If the `-j N` command line option is used, where N > 1, with collection download then the downstream models can simultaneously download and unzip the same upstream model. The result are errors like:

```
[Err] [Zip.cc:161] Error opening: materials/textures/Chainlink_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Chainlink_Transparency.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Door_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Entrance.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/GroundDirt_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Light_Plastic_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Light_Plastic_Emissive.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Pipes_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Rails_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/RockWall_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/RustyMetal_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/Wood_Albedo.jpg
[Err] [Zip.cc:161] Error opening: materials/textures/YellowPlastic_Albedo.jpg
[Err] [Zip.cc:161] Error opening: meshes/
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_01.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_02.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_03.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_04.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_05.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_06.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_07.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_08.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_09.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_Section_10.DAE
[Err] [Zip.cc:161] Error opening: meshes/Edgar_StartingArea.DAE
[Err] [Zip.cc:161] Error opening: model.config
[Err] [Zip.cc:161] Error opening: model.sdf
[Err] [Zip.cc:161] Error opening: thumbnails/
[Err] [Zip.cc:161] Error opening: thumbnails/1.jpg
[Err] [Zip.cc:161] Error opening: thumbnails/2.jpg
[Err] [Zip.cc:161] Error opening: thumbnails/3.jpg
[Err] [Zip.cc:161] Error opening: thumbnails/4.jpg
[Err] [Zip.cc:161] Error opening: thumbnails/5.jpg
[Err] [Zip.cc:118] Source archive does not exist: /home/maspence/.ignition/fuel/fuel.ignitionrobotics.org/openrobotics/models/edgar mine virtual stix/12/edgar mine virtual stix.zip
[Err] [LocalCache.cc:415] Unable to unzip [/home/maspence/.ignition/fuel/fuel.ignitionrobotics.org/openrobotics/models/edgar mine virtual stix/12/edgar mine virtual stix.zip]
```

## Steps to reproduce

These errors can be seen using `ign fuel download -v 4 -j 8 -u "https://fuel.ignitionrobotics.org/openrobotics/collections/SubT Tech Repo"`.

### PR DESCRIPTION
**Pull request by mjcarroll**
# 🎉 New feature

Closes #188 

## Summary

Adds APIs for downloading a set of models in parallel.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: src/FuelClient.cc [hunk 1]
```diff
@@ -719,17 +750,26 @@ Result FuelClient::DownloadModel(const ModelIdentifier &_id,
         std::string dependencyPath;
         ignition::common::URI dependencyURI(meta.dependencies(i).uri());
 
-        // If the model is not already cached, download it; this prevents
-        // any sort of cyclic dependencies from running infinitely
-        if (!this->CachedModel(dependencyURI, dependencyPath))
-          this->DownloadModel(dependencyURI, dependencyPath);
+        ModelIdentifier dependencyID;
+        this->ParseModelUrl(dependencyURI, dependencyID);
```
Comment (jennuine) [inline src/FuelClient.cc:754]: Should we check if parsing failed?

File: src/FuelClient.cc [hunk 2]
```diff
@@ -719,17 +750,26 @@ Result FuelClient::DownloadModel(const ModelIdentifier &_id,
...
+        _dependencies.push_back(dependencyID);
       }
     }
   }
 
   return Result(ResultType::FETCH);
 }
 
+//////////////////////////////////////////////////
+Result FuelClient::DownloadModels(
```
Comment (jennuine) [inline src/FuelClient.cc:764]: Is this function not finished? It doesn't appear to be doing anything and isn't being used

File: include/ignition/fuel_tools/FuelClient.hh [hunk 3]
```diff
@@ -227,6 +227,27 @@ namespace ignition
...
+      /// existing local copy of the model.
+      /// \param[in] _id The model identifier.
+      /// \param[in] _headers Headers to set on the HTTP request.
+      /// \param[out] _dependencies List of models that this model depends on.
+      /// \return Result of the download operation
+      public: Result DownloadModel(const ModelIdentifier &_id,
+                  const std::vector<std::string> &_headers,
+                  std::vector<ModelIdentifier> &_dependencies);
+
+      public: Result DownloadModels(
```
Comment (jennuine) [inline include/ignition/fuel_tools/FuelClient.hh:240]: Missing docs

File: include/ignition/fuel_tools/FuelClient.hh [hunk 4]
```diff
@@ -227,6 +227,27 @@ namespace ignition
...
+                  const std::vector<ModelIdentifier> &_ids,
+                  size_t _jobs = 2);
+
+      /// \brief Retrieve the list of dependencies for a model.
```
Comment (jennuine) [inline include/ignition/fuel_tools/FuelClient.hh:244]: codecheck error: remove line ending whitespace
```suggestion
      /// \brief Retrieve the list of dependencies for a model.
```

File: src/FuelClient.cc [hunk 5]
```diff
@@ -616,6 +616,28 @@ Result FuelClient::DownloadModel(const ModelIdentifier &_id)
...
 Result FuelClient::DownloadModel(const ModelIdentifier &_id,
     const std::vector<std::string> &_headers)
+{
+  std::vector<ModelIdentifier> dependencies;
+  auto res = this->DownloadModel(_id, _headers, dependencies);
+
+  if(!res)
+    return res;
+
+  for (auto dep: dependencies)
```
Comment (jennuine) [inline src/FuelClient.cc:626]: codecheck error
```suggestion
  for (auto dep : dependencies)
```

File: src/FuelClient.cc [hunk 6]
```diff
@@ -678,10 +700,19 @@ Result FuelClient::DownloadModel(const ModelIdentifier &_id,
   if (!this->dataPtr->cache->SaveModel(newId, resp.data, true))
     return Result(ResultType::FETCH_ERROR);
 
+  return this->ModelDependencies(_id, _dependencies);
```
Comment (jennuine) [inline src/FuelClient.cc:703]: codecheck error
```suggestion
  return this->ModelDependencies(_id, _dependencies);
```

**Review by jennuine — COMMENTED** [review]
Looks good so far, appears to be still be in progress. Can this be retarged to Citadel?

**Comment by nkoenig** [comment]
Does this resolve #188  for you? I still see the errors when using a command like this:

```
ign fuel download -v 4 -j 8 --type model -u "https://fuel.ignitionrobotics.org/OpenRobotics/collections/SubT Tech Repo"
```

**Comment by nkoenig** [comment]
@mjcarroll , I think you were missing this line: d16d499af8a136f517f1c2f3c5c0de1386d0b13c? Can you confirm?

**Review by nkoenig — APPROVED** [review]

**Comment by scpeters** [comment]
fix for windows build in #213

File: src/ign.cc [hunk 7]
```diff
@@ -630,93 +630,17 @@ extern "C" IGNITION_FUEL_TOOLS_VISIBLE int downloadUrl(const char *_url,
...
-    }
-
-    // All the tasks have been queued. Now wait for them to finish
-    while (!tasks.empty())
-    {
-      checkForFinishedTasks();
+      auto result = client.DownloadWorlds(worldIds, _jobs);
+      ignerr << "Failed to download worlds for collection ["
+          << collection.Name()
+          << "]" << std::endl;
```
Comment (scpeters) [inline src/ign.cc:643]: is this error message always printed even if the worlds are downloaded successfully?

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-01-10/1228/1

**Pull request merged**